### ISSUE TITLE
question on leveraging data format for learning the following

### ISSUE THREAD

**Issue by jwijffels** [facebookresearch/StarSpace#201]
Hi Starspace authors,

I'm having this type of data where I basically have some plain text. Each text can be labelled with a several categories (multi-label classification, in my data I have more than 500 categories, in the below example I have labels Toolkits, ProgrammingLanguage and NaturalLanguageProcessing. For each of these tagged labels I also have an extra sentiment category which can either be negative/neutral/positive. I would like to build a model for predicting the sentiment category.
Note that in the training data, I don't know which part of the text was taken into account to consider it positive/neutral/negative. 
My question is now how can I leverage the Starspace data setup to make it learn the sentiment for new text.

```
Starspace is a great, comprehensive framework written in C++ for doing NLP    __label__Toolkits                  __label__Positive
Starspace is a great, comprehensive framework written in C++ for doing NLP    __label__ProgrammingLanguage       __label__Neutral
Starspace is a great, comprehensive framework written in C++ for doing NLP    __label__NaturalLanguageProcessing __label__Neutral
```

I tried basically the following approaches in order to allow to extract from which part of the text, the sentiment was about

1. made a multi-label classification model for the categories (tagspace), base on that tagspace model got the cosine distance of each word to each of the >500 labels. Used these similarities between tags and words to construct a new training dataset and used these similarities as weights in a new tagspace model for predicting the sentiment. When I trained that model the loss on the validation set after each epoch went up indicating it wasn't learning what I thought it should be learning
2. made a multi-label classification model for the categories (tagspace), base on that tagspace model got the cosine distance of each word to each of the >500 labels. Looked in the neighbourhood of these words where the similarity to the tag was above a certain threshold and did dependency parsing to see which adjectives in the neighbourhood were of the words. These adjectives were then put as new text data so that I basically had

```
great, comprehensive framework    __label__Toolkits                  __label__Positive
C++                                               __label__ProgrammingLanguage       __label__Neutral
doing NLP                                      __label__NaturalLanguageProcessing __label__Neutral
```
based on that dataset, I build a new tagspace model for the labels positive/neutral/negative
When looking at the evolution of the loss that also jumped up and down for the evaluation dataset for each epoch. Seems like it could not distinguish between the postive/neutral/negative labels
![rplot](https://user-images.githubusercontent.com/1710810/49467825-96006a00-f803-11e8-9919-604d69e80e2a.png)

What can I do with Starspace to approach this problem. I'm using the R package ruimtehol, I'm quite flexible on changing the C++ code to be able to meet the needs. Do you have any suggestions?

**Comment by ledw** [comment]
Hi @jwijffels: what is your goal for this task? Is it predicting sentiment based on new text? If so, how much effect do the category have on that task? Do you also need to predict categories or not? I'd suggest that you treat the category features as words in your task, and use sentiment labels as your labels. What is the size of your train/validation dataset? The sizes of the datasets may explain why you're seeing this graph on validation.

**Comment by jwijffels** [comment]
Goals is twofold
1. predict the categories (already have a tagspace model for that)
2. predict the sentiment, given the category
Data set is about 850000 plain texts, training set was taken to be 90%, validation set was 10% of that.
Category seem to have a big effect as the same text can be positive or negative depending on the category that was assigned to it.
I'll take your advise and use the category features as words in my task. 
Thank you for the help.
